### PR TITLE
Deployment via ModelAPI

### PR DESCRIPTION
**Pull request by akorobeinikov**
TODO:
- [x] Replace Inferencer with models from ModelAPI
- [x] Type of configs and labels
- [x] Add README

Can be merged after: openvinotoolkit/open_model_zoo#2900 and openvinotoolkit/training_extensions#727

### CODE REVIEW

File: mmdet/apis/ote/apis/detection/configuration.py [hunk 1]
```diff
@@ -99,25 +99,61 @@ class __LearningParameters(ParameterGroup):
...
+                                editable=True, visible_in_ui=True)
+
+        @attrs
+        class __Preprocessing(ParameterGroup):
+            header = string_attribute("Preprocessing")
+            description = header
+
+            resize_type = selectable(default_value=ResizeTypes.STANDARD, header="Resize type",
+                                     description="A resize type for model preprocess",
+                                     editable=True, visible_in_ui=True)
```
Comment (druzhkov-paul) [inline mmdet/apis/ote/apis/detection/configuration.py]: Resize type is defined by the model config, not the user.

File: mmdet/apis/ote/apis/detection/configuration.py [hunk 2]
```diff
@@ -99,25 +99,61 @@ class __LearningParameters(ParameterGroup):
...
-            max_value=1,
-            header="Confidence threshold",
-            description="This threshold only takes effect if the threshold is not set based on the result.",
+            header="Using auto parameters",
+            description="Using default parameters for model from ModelAPI",
             affects_outcome_of=ModelLifecycle.INFERENCE
         )
+        class_name = selectable(default_value=Models.SSD, header="Model class for inference",
+                                description="Model classes with defined pre- and postprocessing",
+                                editable=True, visible_in_ui=True)
```
Comment (druzhkov-paul) [inline mmdet/apis/ote/apis/detection/configuration.py]: This should not be the users choice. And it is always "SSD" here.

File: mmdet/apis/ote/apis/detection/configuration.py [hunk 3]
```diff
@@ -99,25 +99,61 @@ class __LearningParameters(ParameterGroup):
...
+            )
+
+            iou_threshold = configurable_float(
+                default_value=0.5,
+                min_value=0,
+                max_value=1,
+                header="Intersection over Union threshold",
+                description="Threshold for NMS detection filtering.",
+                affects_outcome_of=ModelLifecycle.INFERENCE
+            )
```
Comment (druzhkov-paul) [inline mmdet/apis/ote/apis/detection/configuration.py]: NMS is a part of the model and we have no plans for making it twice.

File: mmdet/apis/ote/apis/detection/configuration.py [hunk 4]
```diff
@@ -99,25 +99,61 @@ class __LearningParameters(ParameterGroup):
...
 
     @attrs
-    class __Postprocessing(ParameterGroup):
-        header = string_attribute("Postprocessing")
+    class __InferenceParameters(ParameterGroup):
+        header = string_attribute("Parameters for inference")
         description = header
 
-        result_based_confidence_threshold = configurable_boolean(
+        use_auto_parameters = configurable_boolean(
```
Comment (druzhkov-paul) [inline mmdet/apis/ote/apis/detection/configuration.py]: Don't think that this should be a configurable parameter.

File: mmdet/apis/ote/apis/detection/configuration_enums.py [hunk 5]
```diff
@@ -14,6 +14,23 @@
...
+    SSD = 'ssd'
+    YOLO = 'yolo'
+
+class ResizeTypes(ConfigurableEnum):
+    """
+    This Enum represents the types of resize for preprocessing
+    """
+    STANDARD = 'standard'
+    FIT_TO_WINDOW = 'fit_to_window'
+    FIR_TO_WINDOW_LETTERBOX = 'fit_to_window_letterbox'
```
Comment (druzhkov-paul) [inline mmdet/apis/ote/apis/detection/configuration_enums.py]: FEAR 😱

File: mmdet/apis/ote/apis/detection/openvino_task.py [hunk 6]
```diff
@@ -192,20 +133,18 @@ class OpenVINODetectionTask(IInferenceTask, IEvaluationTask, IOptimizationTask):
...
         self.inferencer = self.load_inferencer()
 
     @property
     def hparams(self):
         return self.task_environment.get_hyper_parameters(OTEDetectionConfig)
 
-    def load_inferencer(self) -> OpenVINODetectionInferencer:
+    def load_inferencer(self):
         labels = self.task_environment.label_schema.get_labels(include_empty=False)
-        self.confidence_threshold = np.frombuffer(self.model.get_data("confidence_threshold"), dtype=np.float32)[0]
```
Comment (druzhkov-paul) [inline mmdet/apis/ote/apis/detection/openvino_task.py:209]: This logic is needed. Confidence threshold is either stored in the model itself, or obtained from the user configuration.

File: mmdet/apis/ote/apis/detection/openvino_task.py [hunk 7]
```diff
@@ -192,20 +133,18 @@ class OpenVINODetectionTask(IInferenceTask, IEvaluationTask, IOptimizationTask):
...
         self.model = self.task_environment.model
-        self.confidence_threshold: float = 0.0
         self.inferencer = self.load_inferencer()
 
     @property
     def hparams(self):
         return self.task_environment.get_hyper_parameters(OTEDetectionConfig)
 
-    def load_inferencer(self) -> OpenVINODetectionInferencer:
+    def load_inferencer(self):
```
Comment (druzhkov-paul) [inline mmdet/apis/ote/apis/detection/openvino_task.py]: What about the return type?

File: mmdet/apis/ote/apis/detection/openvino_task.py [hunk 8]
```diff
@@ -220,10 +159,42 @@ def infer(self, dataset: DatasetEntity, inference_parameters: Optional[Inference
     def evaluate(self,
                  output_result_set: ResultSetEntity,
                  evaluation_metric: Optional[str] = None):
-        if evaluation_metric is not None:
-            logger.warning(f'Requested to use {evaluation_metric} metric, but parameter is ignored. Use F-measure instead.')
```
Comment (druzhkov-paul) [inline mmdet/apis/ote/apis/detection/openvino_task.py]: Do not remove this.

**Review by druzhkov-paul — CHANGES_REQUESTED** [review]

File: mmdet/apis/ote/apis/detection/openvino_task.py [hunk 9]
```diff
@@ -42,139 +49,69 @@
...
-        :param weight_file: Path OpenVINO IR model weights file.
-        :param confidence_threshold: Confidence threshold for passing detection to the output.
+        :param model: Path to model to load, `.xml`, `.bin` or `.onnx` file.
+        :param hparams: Hyper parameters that the model should use.
+        :param num_requests: Maximum number of requests that the inferencer can make.
+            Good value is the number of available cores. Defaults to 1.
         :param device: Device to run inference on, such as CPU, GPU or MYRIAD. Defaults to "CPU".
-        :param num_requests: Maximum number of requests that the inferencer can make. Defaults to 1.
-
         """
```
Comment (druzhkov-paul) [inline mmdet/apis/ote/apis/detection/openvino_task.py]: Please revert the changes.

File: mmdet/apis/ote/apis/detection/openvino_task.py [hunk 10]
```diff
@@ -192,20 +129,21 @@ class OpenVINODetectionTask(IInferenceTask, IEvaluationTask, IOptimizationTask):
...
+            self.hparams.inference_parameters.postprocessing.confidence_threshold = self.confidence_threshold
+        except KeyError:
+            self.confidence_threshold = self.hparams.inference_parameters.postprocessing.confidence_threshold
+        self.model_name = task_environment.model_template.name
         self.inferencer = self.load_inferencer()
 
-    @property
-    def hparams(self):
-        return self.task_environment.get_hyper_parameters(OTEDetectionConfig)
-
```
Comment (druzhkov-paul) [inline mmdet/apis/ote/apis/detection/openvino_task.py]: Just can't understand, what you want to achieve here. Is it necessary to change the logic of the task to make it deployable? If not, please revert the changes related to confidence threshold.

File: mmdet/apis/ote/apis/detection/openvino_task.py [hunk 11]
```diff
@@ -224,6 +162,39 @@ def evaluate(self,
             logger.warning(f'Requested to use {evaluation_metric} metric, but parameter is ignored. Use F-measure instead.')
         output_result_set.performance = MetricsHelper.compute_f_measure(output_result_set).get_performance()
 
+    def deploy(self,
+               output_path: str):
+        work_dir = os.path.dirname(demo.__file__)
+        model_file = inspect.getfile(type(self.inferencer.model))
+        parameters = {}
+        is_new_model = 'model_api' not in model_file
```
Comment (druzhkov-paul) [inline mmdet/apis/ote/apis/detection/openvino_task.py]: What does that mean? Any mentions of model_api are forbidden there?

**Review by druzhkov-paul — CHANGES_REQUESTED** [review]

File: mmdet/apis/ote/apis/detection/openvino_task.py (hunk 10 above)
Comment (akorobeinikov) [reply mmdet/apis/ote/apis/detection/openvino_task.py]: I define confidence threshold in __init__. But for Inferencer I use hparams, because having space for other parameters, not only conf threshold. I revert hparams as read-only property.

Comment (druzhkov-paul) [reply mmdet/apis/ote/apis/detection/openvino_task.py]: hparams has a limited scope -- this is for user interaction mostly. Getting them from task environment guarantee to provide latest values. Please revert.

File: mmdet/apis/ote/apis/detection/openvino_task.py [hunk 12]
```diff
@@ -188,12 +126,14 @@ def __getitem__(self, index):
...
         return len(self.dataset)
 
+
 class OpenVINODetectionTask(IInferenceTask, IEvaluationTask, IOptimizationTask):
     def __init__(self, task_environment: TaskEnvironment):
         logger.info('Loading OpenVINO OTEDetectionTask')
         self.task_environment = task_environment
         self.model = self.task_environment.model
         self.confidence_threshold: float = 0.0
+        self.model_name = task_environment.model_template.name.replace(" ", "_")
```
Comment (druzhkov-paul) [inline mmdet/apis/ote/apis/detection/openvino_task.py]: Please use `task_environment.model_template.model_template_id` instead.

File: mmdet/apis/ote/apis/detection/openvino_task.py [hunk 13]
```diff
@@ -203,11 +143,16 @@ def hparams(self):
 
     def load_inferencer(self) -> OpenVINODetectionInferencer:
         labels = self.task_environment.label_schema.get_labels(include_empty=False)
-        self.confidence_threshold = np.frombuffer(self.model.get_data("confidence_threshold"), dtype=np.float32)[0]
-        return OpenVINODetectionInferencer(labels,
+        _hparams = self.hparams
+        try:
+            self.confidence_threshold = np.frombuffer(self.model.get_data("confidence_threshold"), dtype=np.float32)[0]
+            _hparams.inference_parameters.postprocessing.confidence_threshold = self.confidence_threshold
```
Comment (druzhkov-paul) [inline mmdet/apis/ote/apis/detection/openvino_task.py]: It's better not do so unless you are sure that this is fine from the platform perspective.

File: mmdet/apis/ote/apis/detection/configuration_enums.py [hunk 14]
```diff
@@ -14,6 +14,15 @@
 
 from ote_sdk.configuration import ConfigurableEnum
 
+
+class Models(ConfigurableEnum):
+    """
+    This Enum represents the types of models for inference
+    """
+    SSD = 'ssd'
+    YOLO = 'yolo'
```
Comment (druzhkov-paul) [inline mmdet/apis/ote/apis/detection/configuration_enums.py]: All of the model are "SSDs" here. Please remove that Enum.

File: mmdet/apis/ote/apis/detection/openvino_task.py [hunk 15]
```diff
@@ -41,140 +47,72 @@
...
+        try:
+            model_adapter = OpenvinoAdapter(create_core(), model_file, weight_file, device=device, max_num_requests=num_requests)
+            label_names = [label.name for label in self.labels]
+            self.configuration = {**attr.asdict(hparams.inference_parameters.postprocessing,
+                                  filter=lambda attr, value: attr.name not in ['header', 'description', 'type', 'visible_in_ui']),
+                                  'labels': label_names}
+            self.model = Model.create_model(hparams.inference_parameters.class_name.value, model_adapter, self.configuration)
+            self.model.load()
+        except ValueError as e:
+            print(e)
```
Comment (druzhkov-paul) [inline mmdet/apis/ote/apis/detection/openvino_task.py]: What error is it for? And why is it safe to ignore it?

**Review by druzhkov-paul — CHANGES_REQUESTED** [review]

File: mmdet/apis/ote/apis/detection/openvino_task.py (hunk 12 above)
Comment (akorobeinikov) [reply mmdet/apis/ote/apis/detection/openvino_task.py]: done

File: mmdet/apis/ote/apis/detection/configuration_enums.py (hunk 14 above)
Comment (akorobeinikov) [reply mmdet/apis/ote/apis/detection/configuration_enums.py]: Removed yolo case

File: mmdet/apis/ote/apis/detection/openvino_task.py (hunk 15 above)
Comment (akorobeinikov) [reply mmdet/apis/ote/apis/detection/openvino_task.py]: This block check that model can be found in ModelAPI models or model_wrappers of task. Now ModelAPI doing it itself and I can delete `try except`

File: mmdet/apis/ote/apis/detection/openvino_task.py (hunk 13 above)
Comment (druzhkov-paul) [reply mmdet/apis/ote/apis/detection/openvino_task.py]: I meant that task environment (and hyper parameters as a part of it) should be immutable from with the task. Although this editing might be technically allowed it potentially has some side effects for the outside code. This being said, due to some specifics of implementation this might be not a problem. I suggest to do a copy at least as a workaround.

**Comment by druzhkov-paul** [comment]
Please make sure that deployment works in the clean environment, which is initialized by `init_venv.sh` script.

File: mmdet/apis/ote/apis/detection/configuration.py [hunk 16]
```diff
@@ -90,25 +90,38 @@ class __LearningParameters(ParameterGroup):
...
-        )
+        class_name = selectable(default_value=Models.SSD,
+                                header="Model class for inference",
+                                description="Model classes with defined pre- and postprocessing",
+                                editable=False,
+                                visible_in_ui=True)
+
+        @attrs
+        class __Postprocessing(ParameterGroup):
+            header = string_attribute("Postprocessing")
```
Comment (druzhkov-paul) [inline mmdet/apis/ote/apis/detection/configuration.py]: As you change the structure of configurable parameters here, please also make corresponding adjustments in [tests](https://github.com/openvinotoolkit/mmdetection/blob/dp/nous_sdk_api/tests/test_ote_api.py#L170-171).

**Review by druzhkov-paul — CHANGES_REQUESTED** [review]

**Comment by druzhkov-paul** [comment]
Build 119 🔴

**Comment by akorobeinikov** [comment]
> Please make sure that deployment works in the clean environment, which is initialized by `init_venv.sh` script.

Deployment works, because `init_venv` using ote_sdk and I added modelAPI as requirement for sdk. But if needed I can add modelAPI to mmdet, because it used in mmdet.

File: mmdet/apis/ote/apis/detection/openvino_task.py (hunk 13 above)
Comment (akorobeinikov) [reply mmdet/apis/ote/apis/detection/openvino_task.py]: Ok, using deepcopy

**Comment by druzhkov-paul** [comment]
Yes, please add it to `requirements/runtime.txt`. Thanks.

**Comment by druzhkov-paul** [comment]
Build 126 🔴

**Comment by druzhkov-paul** [comment]
@akorobeinikov Please take a look at the [failed tests](https://ci.iotg.sclab.intel.com/job/ote/job/ote-test/126/).

**Comment by akorobeinikov** [comment]
> @akorobeinikov Please take a look at the [failed tests](https://ci.iotg.sclab.intel.com/job/ote/job/ote-test/126/).

I don't have access to this ci, but I tested locally and get the same result on branches `dp/nous_sdk_api` and `develop` for ote and environment set by `init_venv`. Result: Failed on seventh test and hang or leak on ninth test.

**Comment by druzhkov-paul** [comment]
@akorobeinikov 

> I tested locally and get the same result on branches dp/nous_sdk_api and develop for ote and environment set by init_venv. Result: Failed on seventh test and hang or leak on ninth test.

So, what's the plan?

**Comment by akorobeinikov** [comment]
> @akorobeinikov
> 
> > I tested locally and get the same result on branches dp/nous_sdk_api and develop for ote and environment set by init_venv. Result: Failed on seventh test and hang or leak on ninth test.
> 
> So, what's the plan?

Please run tests on ci to check influence my PR or give me rights to access if you can, that I can take a look at the failed tests.

**Review by druzhkov-paul — APPROVED** [review]
